File: src/Message.cpp
/**
 * @file Message.cpp
 *
 * This module contains the implementation of the Raft::Message class.
 *
 * © 2018 by Richard Walters
 */

#include "Message.hpp"

#include <Json/Value.hpp>
#include <Serialization/SerializedBoolean.hpp>
#include <Serialization/SerializedInteger.hpp>
#include <Serialization/SerializedString.hpp>
#include <Serialization/SerializedUnsignedInteger.hpp>
#include <Serialization/SerializedVector.hpp>
#include <SystemAbstractions/StringFile.hpp>

namespace {

    constexpr int CURRENT_SERIALIZATION_VERSION = 2;

}

namespace Raft {

    Message::Message(const std::string& serialization) {
        SystemAbstractions::StringFile buffer(serialization);
        Serialization::SerializedUnsignedInteger version;
        if (!version.Deserialize(&buffer)) {
            return;
        }
        if (version > CURRENT_SERIALIZATION_VERSION) {
            return;
        }
        Serialization::SerializedInteger serializedType;
        if (!serializedType.Deserialize(&buffer)) {
            return;
        }
        type = (Message::Type)(int)serializedType;
        Serialization::SerializedInteger intField;
        if (!intField.Deserialize(&buffer)) {
            return;
        }
        term = intField;
        if (version >= 2) {
            if (!intField.Deserialize(&buffer)) {
                return;
            }
            seq = intField;
        }
        switch (type) {
            case Message::Type::RequestVote: {
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                requestVote.candidateId = intField;
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                requestVote.lastLogIndex = (size_t)intField;
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                requestVote.lastLogTerm = intField;
            } break;

            case Message::Type::RequestVoteResults: {
                Serialization::SerializedBoolean boolField;
                if (!boolField.Deserialize(&buffer)) {
                    return;
                }
                requestVoteResults.voteGranted = boolField;
            } break;

            case Message::Type::AppendEntries: {
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                appendEntries.leaderCommit = (size_t)intField;
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                appendEntries.prevLogIndex = (size_t)intField;
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                appendEntries.prevLogTerm = intField;
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                const auto numLogEntries = (size_t)intField;
                log.reserve(numLogEntries);
                for (size_t i = 0; i < numLogEntries; ++i) {
                    LogEntry logEntry;
                    if (!logEntry.Deserialize(&buffer)) {
                        return;
                    }
                    log.push_back(std::move(logEntry));
                }
            } break;

            case Message::Type::AppendEntriesResults: {
                Serialization::SerializedBoolean boolField;
                if (!boolField.Deserialize(&buffer)) {
                    return;
                }
                appendEntriesResults.success = boolField;
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                appendEntriesResults.matchIndex = (size_t)intField;
            } break;

            case Message::Type::InstallSnapshot: {
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                installSnapshot.lastIncludedIndex = (size_t)intField;
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                installSnapshot.lastIncludedTerm = intField;
                Serialization::SerializedString stringField;
                if (!stringField.Deserialize(&buffer)) {
                    return;
                }
                snapshot = Json::Value::FromEncoding(stringField);
            } break;

            case Message::Type::InstallSnapshotResults: {
                if (!intField.Deserialize(&buffer)) {
                    return;
                }
                installSnapshotResults.matchIndex = (size_t)intField;
            } break;

            default: return;
        }
    }

    std::string Message::Serialize() const {
        SystemAbstractions::StringFile buffer;
        Serialization::SerializedInteger version(CURRENT_SERIALIZATION_VERSION);
        if (!version.Serialize(&buffer)) {
            return "";
        }
        Serialization::SerializedInteger intField((int)type);
        if (!intField.Serialize(&buffer)) {
            return "";
        }
        intField = term;
        if (!intField.Serialize(&buffer)) {
            return "";
        }
        intField = seq;
        if (!intField.Serialize(&buffer)) {
            return "";
        }
        switch (type) {
            case Message::Type::RequestVote: {
                intField = requestVote.candidateId;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                intField = (int)requestVote.lastLogIndex;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                intField = requestVote.lastLogTerm;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
            } break;

            case Message::Type::RequestVoteResults: {
                Serialization::SerializedInteger boolField(requestVoteResults.voteGranted);
                if (!boolField.Serialize(&buffer)) {
                    return "";
                }
            } break;

            case Message::Type::AppendEntries: {
                intField = (int)appendEntries.leaderCommit;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                intField = (int)appendEntries.prevLogIndex;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                intField = appendEntries.prevLogTerm;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                intField = (int)log.size();
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                for (const auto& logEntry: log) {
                    if (!logEntry.Serialize(&buffer)) {
                        return "";
                    }
                }
            } break;

            case Message::Type::AppendEntriesResults: {
                Serialization::SerializedInteger boolField(appendEntriesResults.success);
                if (!boolField.Serialize(&buffer)) {
                    return "";
                }
                intField = (int)appendEntriesResults.matchIndex;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
            } break;

            case Message::Type::InstallSnapshot: {
                intField = (int)installSnapshot.lastIncludedIndex;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                intField = installSnapshot.lastIncludedTerm;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
                Serialization::SerializedString stringField(snapshot.ToEncoding());
                if (!stringField.Serialize(&buffer)) {
                    return "";
                }
            } break;

            case Message::Type::InstallSnapshotResults: {
                intField = (int)installSnapshotResults.matchIndex;
                if (!intField.Serialize(&buffer)) {
                    return "";
                }
            } break;

            default: return "";
        }
        return buffer;
    }

}

File: src/message.rs
use crate::LogEntry;
use serde::{
    Deserialize,
    Serialize,
};
use std::hash::Hash;

/// This holds the information contained in an [`AppendEntries`]
/// message from one server to another.
///
/// [`AppendEntries`]: enum.MessageContent.html#variant.AppendEntries
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct AppendEntriesContent<T, Id>
where
    Id: Eq + Hash,
{
    /// This is the index of the last log entry which the leader of
    /// the cluster has committed at the time this message was sent.
    pub leader_commit: usize,

    /// This is the index of the log entry immediately before the first
    /// log entry in this message.  The index of the first log entry is 1,
    /// so this will be 0 if this message holds the first entry in the log.
    pub prev_log_index: usize,

    /// This is the cluster leadership term that was in effect when the log
    /// entry before the first log entry in this message was created.
    pub prev_log_term: usize,

    /// This holds the entries to be appended to the log.
    pub log: Vec<LogEntry<T, Id>>,
}

/// This holds the content of a message sent from one server to another.
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum Content<S, T, Id>
where
    Id: Eq + Hash,
{
    /// The sender is running for cluster leadership and is requesting
    /// the receiver vote for the sender.
    RequestVote {
        /// This is the index of the last log entry the candidate for
        /// leadership has in its copy of the log.
        last_log_index: usize,

        /// This is the cluster leadership term that was in effect when the
        /// last log entry the candidate for leadership has in its copy
        /// of the log was created.
        last_log_term: usize,
    },

    /// The sender is responding to a previous [`RequestVote`] sent by the
    /// receiver.
    ///
    /// [`RequestVote`]: #variant.RequestVote
    RequestVoteResponse {
        /// This indicates whether or not the sender is voting for the
        /// receiver.
        vote_granted: bool,
    },

    /// Either append new entries to the receiver's copy of the log,
    /// or (if no log entries are provided) serve as a "heartbeat" to prevent
    /// the receiver from starting a new election.
    AppendEntries(AppendEntriesContent<T, Id>),

    /// The sender is responding to a previous [`AppendEntries`] sent by
    /// the receiver.
    ///
    /// [`AppendEntries`]: #variant.AppendEntries
    AppendEntriesResponse {
        /// Indicate whether or not the new log entries were accepted.
        success: bool,

        /// Indicate the index of the next log entry the receiver should
        /// send back.
        next_log_index: usize,
    },

    /// Replace the server state and compacted log with a given snapshot.
    InstallSnapshot {
        /// This is the index of the last log entry incorporated into the
        /// given snapshot.
        last_included_index: usize,

        /// This is the cluster leadership term that was in effect when the
        /// last log entry incorporated into the given snapshot was created.
        last_included_term: usize,

        /// This is the complete server state and compacted log that
        /// should be accepted by the receiver.
        snapshot: S,
    },

    /// The sender is responding to a previous [`InstallSnapshot`] sent by
    /// the receiver, indicating that it's ready to accept more log entries
    /// based on the received snapshot.
    ///
    /// [`InstallSnapshot`]: #variant.InstallSnapshot
    InstallSnapshotResponse {
        /// Indicate the index of the next log entry the receiver should
        /// send back.
        next_log_index: usize,
    },
}

/// This holds information to be sent from one server to another.
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct Message<S, T, Id>
where
    Id: Eq + Hash,
{
    /// This holds information which varies according to the intention
    /// of the message.
    pub content: Content<S, T, Id>,

    /// This is a number used to correlate request messages with responses.
    /// It should be different for each separate request sent from the
    /// cluster leader to any specific follower.
    pub seq: usize,

    /// This is the current cluster leadership term in effect for the
    /// sender of the message, used to inform the receiver in order to
    /// synchronize the servers to be in the same term.
    pub term: usize,
}
